ethcore/evm/src/interpreter/mod.rs

```diff
@@ -102,127 +103,234 @@ enum InstructionResult<Gas> {
-					let pos = self.verify_jump(position, jump_destinations)?;
-					reader.position = pos;
-				},
-				InstructionResult::StopExecutionNeedsReturn {gas, init_off, init_size, apply} => {
-					informant.done();
-					let mem = mem::replace(&mut self.mem, Vec::new());
-					return Ok(GasLeft::NeedsReturn {
-						gas_left: gas.as_u256(),
-						data: mem.into_return_data(init_off, init_size),
-						apply_state: apply
-					});
-				},
-				InstructionResult::StopExecution => break,
-				_ => {},
-			}
-		}
-		informant.done();
-		Ok(GasLeft::Known(gasometer.current_gas.as_u256()))
-	}
-}
+				if self.do_trace {
+					ext.trace_executed(
+						self.gasometer.current_gas.as_u256(),
+						self.stack.peek_top(info.ret),
+						mem_written.map(|(o, s)| (o, &(self.mem[o..o+s]))),
+						store_written,
+					);
+				}
 
-impl<Cost: CostType> Interpreter<Cost> {
-	/// Create a new `Interpreter` instance with shared cache.
-	pub fn new(cache: Arc<SharedCache>) -> Interpreter<Cost> {
-		Interpreter {
-			mem: Vec::new(),
-			cache: cache,
-			return_data: ReturnData::empty(),
-			_type: PhantomData::default(),
+				// Advance
+				match result {
+					InstructionResult::JumpToPosition(position) => {
+						if self.valid_jump_destinations.is_none() {
+							let code_hash = self.params.code_hash.clone().unwrap_or_else(|| keccak(self.reader.code.as_ref()));
+							self.valid_jump_destinations = Some(self.cache.jump_destinations(&code_hash, &self.reader.code));
+						}
+						let jump_destinations = self.valid_jump_destinations.as_ref().expect("jump_destinations are initialized on first jump; qed");
+						let pos = try_or_done!(self.verify_jump(position, jump_destinations));
+						self.reader.position = pos;
+					},
+					InstructionResult::StopExecutionNeedsReturn {gas, init_off, init_size, apply} => {
+						let mem = mem::replace(&mut self.mem, Vec::new());
+						return InterpreterResult::Done(Ok(GasLeft::NeedsReturn {
+							gas_left: gas.as_u256(),
+							data: mem.into_return_data(init_off, init_size),
+							apply_state: apply
+						}));
+					},
+					InstructionResult::StopExecution => {
+						return InterpreterResult::Done(Ok(GasLeft::Known(self.gasometer.current_gas.as_u256())));
+					},
+					_ => {},
+				}
+
+				if self.reader.position >= self.reader.len() {
+					return InterpreterResult::Done(Ok(GasLeft::Known(self.gasometer.current_gas.as_u256())));
+				}
+
+				InterpreterResult::Continue
+			};
+			inner()
+		};
+		if let &InterpreterResult::Done(_) = &result {
+			self.done = true;
+			self.informant.done();
 		}
+		return result;
 	}
 
-	fn verify_instruction(&self, ext: &vm::Ext, instruction: Instruction, info: &InstructionInfo, stack: &Stack<U256>) -> vm::Result<()> {
+	fn verify_instruction(&self, ext: &vm::Ext, instruction: Instruction, info: &InstructionInfo) -> vm::Result<()> {
 		let schedule = ext.schedule();
```

Comment:
Schedule could probably be cached on `self` level, less virtual calls.

Reply:
The issue is that for some use cases of `step`, like pausing, we would want to take back the `Ext` reference, possibly do some modifications to it, before passing it back to continue the execution.

1. If we take a reference of `Schedule` and put it on self level, it means that this single `Ext` is not usable again for the whole lifecycle of the Interpreter.
2. Or we need to clone the `Schedule`, which may also be expensive.

I think virtual calls might not be *that* expensive. And with (1), one optimization we can do is to split `Ext` up to a) interpreter-specific `Ext`, and b) transaction-level `Ext`. I think that may be a good trade-off compared with additional virtual calls during each `step`.

Reply:
Makes sense. I thought that we already clone schedule on `ext.schedule()` but now saw it's only borrowed.